Testing unstable ssa values useref values in red

joseeloren opened the pull request:
I have made some tests to check if unstable ssa values useref values are in red. This address issue #54028 PRs: #54501 and #54441. The last one is #54501, that is waiting for these tests.

@gbaraldi

joseeloren commented:
@gbaraldi @Zentrik @fingolfin

gbaraldi commented:
I think it's almost there but looking at it in my terminal it seems we are printing the SSA values in a slightly different colour
![image](https://github.com/user-attachments/assets/1cb24a23-c720-4d07-a588-6649a5eea3e0)

joseeloren commented:
You are right @gbaraldi Fixed.

lgoettgens commented on base/show.jl:
```diff
@@ -1,6 +1,7 @@
 
-using Core.Compiler: has_typevar
+using Core.Compiler: has_typevar, userefs, scan_ssa_use!
+import Base: Base
```
```suggestion
```
This seems unnecessary

joseeloren replied:
Right. Removed.

lgoettgens reviewed (approved):
The output now seems fine to me. Furthermore, all comments in https://github.com/JuliaLang/julia/pull/54501 seem to be adapted (already by @arhik in https://github.com/JuliaLang/julia/pull/54501).

To whoever is merging this: please close https://github.com/JuliaLang/julia/issues/54028 and https://github.com/JuliaLang/julia/pull/54501 at the same time

topolarity commented on base/compiler/ssair/show.jl:
```diff
@@ -31,11 +31,17 @@ function Base.show(io::IO, cfg::CFG)
 
-function print_stmt(io::IO, idx::Int, @nospecialize(stmt), used::BitSet, maxlength_idx::Int, color::Bool, show_type::Bool)
+function print_stmt(io::IO, idx::Int, @nospecialize(stmt), used::BitSet, maxlength_idx::Int, color::Bool, show_type::Bool,
+                    unstable_ssa::Union{Nothing, BitSet} = nothing)
```
Why the `nothing` case instead of `BitSet()`, or requiring this parameter always?

lgoettgens replied:
I think this was changed due to @fingolfin's comment https://github.com/JuliaLang/julia/pull/54501#discussion_r1612384199

topolarity replied:
Hmm.. on second thought, I think this should probably not tie unstable-highlighting to `show_type` in the first place

I think we want, e.g., `π(%26)` and `return %10` to show the type-unstable SSA value in red:
![image](https://github.com/user-attachments/assets/f1dc14b0-a0af-4da0-93c9-0780a135b248)

joseeloren replied:
Maybe we need @fingolfin opinion to continue?

fingolfin replied:
Hi folks, I never saw the pings for me here (I get dozens and sometimes hundreds of GitHub notifications per day and it is very easy to miss something). If you want to make sure I look at something, I recommend pinging me on Slack or sending me an email.

That said, I am not even sure what I am being asked here. Looking at PR #54501 I made no comments about what parts to show in which color I think? The comment @lgoettgens links to is about a refactoring of the code that should be neutral with respect to behavior. The goal was to avoid unnecessary allocations and code duplication. The suggestion by @topolarity addresses the first point, and the second is not that important. But in any case, I agree with @topolarity that none of this is a reason to hold up the changes (I thought that I already made that clear in my comments on #54501, by stating "Left some minor and optional suggestions, feel free to ignore/disagree" ... :-) )

topolarity commented on base/compiler/ssair/show.jl:
```diff
@@ -31,11 +31,17 @@ function Base.show(io::IO, cfg::CFG)
         pad = " "^(maxlength_idx - length(idx_s) + 1)
-        print(io, "%", idx_s, pad, "= ")
+        if unstable_ssa != nothing && idx in unstable_ssa
+            printstyled(io, "%", idx_s; color=:light_red, bold=true)
```
Does this still print correctly for green (i.e. new) nodes?

Seems like it might not be respecting the `color::Bool` argument

joseeloren replied:
AFAIK, color was not used previously. I think we can remove color::Bool argument

topolarity replied:
? It's used just below this on line 44/50

joseeloren replied:
Sorry, I did'nt explain it correctly. I mean that color is only check there, but if some color is defined, it is not use in any function as an argument, IINM

joseeloren replied:
I think it is solved, if I understood the issue.

joseeloren replied:
@topolarity

joseeloren replied:
Good catch @topolarity, you were completely right.

I have fixed this in the latest rebase. The highlight check inside print_stmt now explicitly gates the ANSI styling behind the color flag:

``` julia
if color && unstable_ssa isa BitSet && idx in unstable_ssa
    printstyled(io, "%", idx_s; color=:light_red, bold=true)
else
    print(io, "%", idx_s)
end
```

Furthermore, I've added a comprehensive test suite in test/show.jl that explicitly tests both color=true (expecting the red \e[91m ANSI code) and color=false (verifying the complete absence of ANSI escape codes while still printing the strict SSA value representation).

It now guarantees graceful degradation for environments without color support!

lgoettgens commented:
Bump @topolarity @fingolfin @gbaraldi . I think this is just waiting for some reaction from your side to the authors changes and responses to your comments in this PR and in https://github.com/JuliaLang/julia/pull/54501.

And it would be great to get this in 1.12 (aka merged until tomorrow)

fingolfin commented:
The main obstacle here seems to be the conflicts this PR has which somebody needs to resolve before merging is an option?

joseeloren committed:
Highlight unstable SSA values in light_red+bold in @code_warntype output

Issue #54028: When displaying code via @code_warntype, SSA values with
unstable types (type == Any) are now highlighted in light_red with bold
formatting. This makes it easier to visually identify type instabilities.

Uses IOContext to pass unstable_ssa information through the IO stream,
avoiding parameter threading through dozens of show functions.

Changes:
- Compiler/src/ssair/show.jl: print_stmt reads :unstable_ssa from IOContext.
  show_ir(CodeInfo) computes unstable SSA set and stores in IOContext.
  Added is_unstable_ssa helper.
- base/show.jl: show_unquoted(SSAValue) and show_unquoted(SlotNumber)
  read :unstable_ssa from IOContext for conditional highlighting.
- test/show.jl: Added tests for unstable SSA highlighting and PiNode color.

Addresses review comments:
- show_type no longer gates unstable highlighting
- color::Bool is respected for unstable SSA formatting
- Uses IOContext instead of parameter threading (topolarity suggestion)
- EMPTY_BITSET pattern subsumed by IOContext approach

joseeloren committed:
Merge branch 'master' into testing_unstable_SSA_values_useref_values_in_red

joseeloren committed:
Merge branch 'master' into testing_unstable_SSA_values_useref_values_in_red

joseeloren committed:
Merge branch 'master' into testing_unstable_SSA_values_useref_values_in_red

joseeloren committed:
Merge branch 'master' into testing_unstable_SSA_values_useref_values_in_red

lgoettgens commented on test/show.jl:
```diff
@@ -2915,3 +2915,98 @@ let m = only(methods(f_show_method))
+        # No light_red (91) highlighting for stable SSA values
+        @test !contains(str, "\e[91m\e[1m%")
+    end
+
+    # Test 4: Body type annotation shows ::Any for unstable
+    @testset "body type ::Any shown for unstable function" begin
+        io = IOBuffer()
+        ioc = IOContext(io, :color => true)
+        code_warntype(ioc, foo_unstable_ssa, (Vector{Any},), optimize=true)
+        str = String(take!(io))
+
+        # The return type should indicate instability with ::Any
+        @test contains(str, "Any")
+    end
```
I am not sure what this testset is supposed to test. Already previous to this PR, we would expect `Body::Any` to be included in the output (with some coloring of course), and this PR does not change anything about that.
Furthermore, this test only tests for the presence of `Any` anywhere in the output, but not in the same line as `Body::`. I would thus suggest to just drop this testset

```suggestion
```

lgoettgens commented on Compiler/src/ssair/show.jl:
```diff
@@ -965,6 +971,14 @@ function show_ir(io::IO, ir::IRCode, config::IRShowConfig=default_config(io, ir)
+    types isa Vector{Any} || return false
+    isassigned(types, idx) || return false
+    type = types[idx]
+    return type === Any
```
This check does not work, as not only `Any` is a unstable ssa type, see e.g.

<img width="597" height="283" alt="Image" src="https://github.com/user-attachments/assets/3269c8a3-1fbd-41b7-9902-671ebc03aee4" />(%3 should also be shown in red).
I think it would be better to use the same decision as for the trailing type annotations (from `warntype_type_printer` in https://github.com/JuliaLang/julia/blob/2325e6a6fdea767b34f6d10b7f4940baa9284829/stdlib/InteractiveUtils/src/codeview.jl#L52, or even better refactor that into its own helper function that can be used in both places.

This would then also go towards also having yellow ssa values like in 

<img width="547" height="821" alt="Image" src="https://github.com/user-attachments/assets/5cfe5bf7-0017-4871-8e54-9cc7eda05efc" />

joseeloren replied:
@lgoettgens Sure, I will appreciate it if you could give a try with Codex. Currently, I am a bit busy.

lgoettgens reviewed (changes requested):
Some more comments (after playing around with it). 
@joseeloren If you currently don't have the capacity for this, I could also give this a try (using Codex). Just let me know :)

joseeloren commented:
> Some more comments (after playing around with it). @joseeloren If you currently don't have the capacity for this, I could also give this a try (using Codex). Just let me know :)

@lgoettgens Sure, I will appreciate it if you could give a try with Codex. Currently, I am a bit busy.

lgoettgens commented:
> > Some more comments (after playing around with it). @joseeloren If you currently don't have the capacity for this, I could also give this a try (using Codex). Just let me know :)
> 
> @lgoettgens Sure, I will appreciate it if you could give a try with Codex. Currently, I am a bit busy.

See https://github.com/JuliaLang/julia/pull/61879.
